### ISSUE TITLE
TypeError: No method has been defined for writing <class 'pandas.core.arrays.string_.StringArray'> elements to <class 'zarr.hierarchy.Group'>

### ISSUE THREAD

**Issue by slowkow** [scverse/anndata#925]
# Reproducible Example

This code fails with an error:

```python
import anndata
import zarr
from scipy import sparse
import numpy as np
import pandas as pd

X = sparse.csr_matrix(sparse.random(m=100, n=1000, density=0.05, random_state=42))

obs = pd.DataFrame({'cell': np.arange(100), 'cat': ['a','b','c','d'] * 25})

# This line causes the error
obs = obs.convert_dtypes()

ad = anndata.AnnData(X=X, obs=obs, dtype=np.float32)
ad.write_zarr('mytest')
```
```
File ~/miniconda3/lib/python3.10/site-packages/anndata/_io/utils.py:220, in report_write_key_on_error.<locals>.func_wrapper(elem, key, val, *args, **kwargs)
    218 else:
    219     parent = _get_parent(elem)
--> 220     raise type(e)(
    221         f"{e}\n\n"
    222         f"Above error raised while writing key {key!r} of {type(elem)} "
    223         f"to {parent}"
    224     ) from e

TypeError: No method has been defined for writing <class 'pandas.core.arrays.string_.StringArray'> elements to <class 'zarr.hierarchy.Group'>

Above error raised while writing key 'categories' of <class 'zarr.hierarchy.Group'> to <zarr.storage.DirectoryStore object at 0x7f18b580c670>
```

This line causes the error:

```python
obs = obs.convert_dtypes()
```

If we comment out this line, then the code runs without the error.

## Details

First, let's have a look at the object that causes the error:

```python
obs = pd.DataFrame({'cell': np.arange(100), 'cat': ['a','b','c','d'] * 25})
obs = obs.convert_dtypes()
obs.cat
```
```
0     a
1     b
2     c
3     d
4     a
     ..
95    d
96    a
97    b
98    c
99    d
Name: cat, Length: 100, dtype: string
```

Next, let's look at the object that works without an error:

```python
obs = pd.DataFrame({'cell': np.arange(100), 'cat': ['a','b','c','d'] * 25})
obs.cat
```
```
0     a
1     b
2     c
3     d
4     a
     ..
95    d
96    a
97    b
98    c
99    d
Name: cat, Length: 100, dtype: object
```

The code that throws the error has `dtype: string`

The code that runs without error has `dtype: object`

Honestly, I don't understand the difference or why anndata cares about it.

I wanted to report it here, just in case someone else out there runs into the same issue. It took me a few hours to distill the code down to a minimal example so I could debug this, and I don't want others to waste their time on this.

## pip list

<details>
<summary>pip list</summary>
<pre>
Package                 Version
----------------------- ---------
anndata                 0.8.0
asciitree               0.3.3
asttokens               2.2.1
backcall                0.2.0
brotlipy                0.7.0
certifi                 2022.12.7
cffi                    1.15.1
charset-normalizer      2.0.4
conda                   23.1.0
conda-content-trust     0.1.3
conda-package-handling  2.0.2
conda_package_streaming 0.7.0
cryptography            38.0.4
decorator               5.1.1
entrypoints             0.4
executing               1.2.0
fasteners               0.18
h5py                    3.8.0
idna                    3.4
ipython                 8.10.0
jedi                    0.18.2
matplotlib-inline       0.1.6
natsort                 8.2.0
numcodecs               0.11.0
numpy                   1.24.2
packaging               23.0
pandas                  1.5.3
parso                   0.8.3
pexpect                 4.8.0
pickleshare             0.7.5
pip                     22.3.1
pluggy                  1.0.0
prompt-toolkit          3.0.37
ptyprocess              0.7.0
pure-eval               0.2.2
pycosat                 0.6.4
pycparser               2.21
Pygments                2.14.0
pyOpenSSL               22.0.0
PySocks                 1.7.1
python-dateutil         2.8.2
pytz                    2022.7.1
requests                2.28.1
ruamel.yaml             0.17.21
ruamel.yaml.clib        0.2.6
scipy                   1.10.1
setuptools              65.6.3
six                     1.16.0
stack-data              0.6.2
toolz                   0.12.0
tqdm                    4.64.1
traitlets               5.9.0
urllib3                 1.26.14
wcwidth                 0.2.6
wheel                   0.37.1
zarr                    2.14.2
zstandard               0.18.0

</pre>
</details>

**Comment by slowkow** [comment]
Thanks for asking!

My old script used anndata with `write_zarr()` and it worked with `convert_dtypes()`. But when I updated to a new python environment on a new server, the script failed. So, I thought I might share my experience here.

In my old script, in order to get `write_zarr()` working I had to manually change integer columns from 'Int64' to 'Int32', and manually convert categorical columns from 'object' to 'category'. In that version of the code, it seemed to me at the time that `convert_dtypes()` was necessary to get `write_zarr()` to work. (I could be mistaken.)
